Add production flags to dataone repositories

ThomasThelen opened the pull request:
We need to tell whether we're publishing to production or dev DataONE. This PR 

1. Sends the user to prod/dev DataONE login
2. Uses the right prod/dev JWT
3. Lets the backend know whether it's prod/dev

The biggest UI change is that the user is prompted to login to DataONE _after_ they select a repository. The reason for this is that dev and production dataone uses two different login servers (and we need to know which one to send the user to).
To Test:
See [this issue](https://github.com/whole-tale/gwvolman/pull/72)

Note: I accidentally published a few Tales to production using this, so be careful when playing with the repository dropdown.

bodom0015 commented:
> Note: I accidentally published a few Tales to production using this, so be careful when playing with the repository dropdown.

Do DataONE's production repositories allow you to delete or archive published artifacts (or somehow hide them from search)? I'm happy to test out the dropdown presentation and everything, but am struggling to find a way to test the functionality here without polluting their production resources with dummy files.

Any advice on how best to test this?

craig-willis commented:
@ThomasThelen  As discussed on the dev call today, the recommendation for v0.7 is to replace the `isProduction` flag with the CN URL for each repository and pass it through instead.  This should mean that you can drop the hardcoded CNs and CN check logic in `getDataONEJWT` and `getEndpoint`.  

@bodom0015  Tommy and I agreed that we would publish a legitimate Tale as our test.  This will likely be some variation of the "Water Tale" that we could reasonably see ourselves referencing via DOI.  It is possible for a DataONE administrator to delete the package after-the-fact, but not recommended.

Other notes:
* Q. Should production users be able to publish to dev?  This is apparently recommended by DataONE for users to test the publication project without polluting the production repository (similar to Zenodo sandbox).  This seems like a reasonable capability to support.  However, we do not want dev instances to support publishing to production.

ThomasThelen commented:
I've made the appropriate changes to send the coordinating node to the backend in ca4c565459d38c532ac19e2557d4569ce1c31a20.
In the process you'll see that I renamed a few variables to make it more clear which node we're dealing with. I also renamed `getEndpoint` to `getPortalEndpoint`.

craig-willis commented on app/components/dashboard/run/left-panel/component.js:
```diff
@@ -361,38 +340,44 @@ export default Component.extend(FullScreenMixin, {
             const self = this;
             console.log('Now publishing:', tale);
+
+            let targetRepo = self.get('selectedRepositoryName');
```
Very minor observation, but couldn't you just use the selected index instead of getting the name and using a separate function to lookup by name?

ThomasThelen replied:
The semantic ui dropdown doesn't have a method for retrieving the index, but it does for text https://semantic-ui.com/modules/dropdown.html#/usage Scroll down to `Behavior`.
I use `get text` [here](https://github.com/whole-tale/dashboard/blob/ca4c565459d38c532ac19e2557d4569ce1c31a20/app/components/dashboard/run/left-panel/component.js#L405), every time the repository is changed. I made [getRepositoryFromName](https://github.com/whole-tale/dashboard/blob/ca4c565459d38c532ac19e2557d4569ce1c31a20/app/components/dashboard/run/left-panel/component.js#L219) as a way to get from the name, to the repository.


In 366 I could have used self.get('selectedRepository') and saved the call on 367 since it should be set by the time this code is reached.

bodom0015 replied:
You can actually use the built-in prototype method `Array.find()` for this too.

`let selectedRepo = repositories.find((repo) => repo.name === selectedRepositoryName);` would do the same thing without needing to write/maintain a custom helper function for it.

craig-willis commented:
@ThomasThelen if we don't want developers to publish to DataONE production, then we need some sort of configuration to determine whether to display the prod option, right?

craig-willis commented on app/components/dashboard/run/left-panel/component.js:
```diff
@@ -361,38 +340,44 @@ export default Component.extend(FullScreenMixin, {
-            }
-            
-            this.set('dataoneJWT', jwt);
+        authenticateD1(taleId, coordinatingNode) {
```
I may be misunderstanding something, but `authenticateD1` seems to be called before I select the member node to publish to.  This means that the coordinatingNode is always the staging node, since the dropdown defaults to the first entry (dev). As discussed on the dev call, this would work since the JWT is raw ORCID and identical for stage/prod. But, for this to work as intended it seems that the user would need to select the member node before authenticating, right?

ThomasThelen replied:
`authenticateD1` gets called when the user [clicks Continue to DataOne Login](https://github.com/whole-tale/dashboard/blob/ca4c565459d38c532ac19e2557d4569ce1c31a20/app/components/dashboard/run/left-panel/template.hbs#L102) after checking the token endpoint. I took out the call to the publish endpoint and confirmed that the correct nodes are checked on development and on production.

In the link above, `selectedRepository` is set each time the repository is changed, and should be current when this code is hit.

craig-willis replied:
Clearly user error on my part -- maybe a stale dashboard build.  I see now that the login prompt appears after I hit "Publish" (after selecting the target node) and that the correct CN is checked for the token.  During my testing, the login prompt was appearing when I clicked the "Publish Tale..." menu, but that's changed.

ThomasThelen commented:
> @ThomasThelen if we don't want developers to publish to DataONE production, then we need some sort of configuration to determine whether to display the prod option, right?

We can hide it if we want by using the dev flag in the environment, but I think everyone knows to not publish to production. It might actually be nice to have a confirmation popup, warning that you won't be able to delete the published Tale.

craig-willis commented:
I'd like for @mbjones  to weigh on whether to disable publish to production in our dev and staging instances.  I disagree that everyone knows not to.  Developers, PIs and other stakeholders tinkering with dev or staging instances will not necessarily know that they shouldn't.

bodom0015 commented on app/components/dashboard/run/left-panel/template.hbs:
```diff
@@ -126,7 +126,7 @@
           <div class="required inline field">
               <label>Please choose a target repository:</label>
-              {{#ui-dropdown class=repoDropdownClass allowAdditions=false selected=selectedRepository onChange=(action 'onRepositoryChange')
+              {{#ui-dropdown class=repoDropdownClass allowAdditions=false selected=selectedRepositoryName onChange=(action 'onRepositoryChange')
```
Forgive me, but I'm not understanding the changes from ` selectedRepository` to `selectedRepositoryName` here.

If we used to select the entire repository object with the dropdown, then why change it to select only the name if we would then need to immediately do a lookup by name to get the full object? Am I just overlooking something obvious?

mbjones commented:
Yes, please do not allow publication to production from dev environments.  People will mistakenly publish things there.  We've already had to manage a few deletions from Tommy.  Can't the list of endpoints be a configuration setting and set up differently for each environment?

ThomasThelen commented:
In regards to 035fd7f, `ENV.dev` was getting set to 'true' or 'false', but as strings instead of boolean type. That commit will turn the value into a bool, which was causing trouble in 
8012834

bodom0015 commented on app/components/dashboard/run/left-panel/component.js:
```diff
@@ -110,38 +109,30 @@ export default Component.extend(FullScreenMixin, {
+                reload: true
+              })
+              .then(resp => {
+              this.send('openPublishModal', resp)
```
Tale should already be available via the `this.model.tale` (and will be `this.model` after the 0.8 refactoring). Was there a reason to requery for it here?

bodom0015 commented on app/components/dashboard/run/left-panel/component.js:
```diff
@@ -110,38 +109,30 @@ export default Component.extend(FullScreenMixin, {
+              })
+              .then(resp => {
+              this.send('openPublishModal', resp)
+            })
```
Tale should already be available via the `this.model.tale` (and will be `this.model` after the 0.8 refactoring). Is there a reason to requery for it here?

ThomasThelen replied:
Since `this.model` is the instance, I'm pretty sure it only has the ID of the Tale. To make sure, I tried to log `this.model.tale` just before the request and I got `undefined`

bodom0015 replied:
Ahh, my apologies... the Run controller grabs all of this data, but it appears to be missing a line [here](https://github.com/whole-tale/dashboard/blob/master/app/controllers/run.js#L14) to actually save the Tale model that we fetch.

In the refactored Run view this should be available via the model, so I can address tidying up on that branch.
